File: c/circularqueue.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_SIZE 100

typedef struct {
  int items[MAX_SIZE];
  int front;
  int rear;
} CircularQueue;


void init_queue(CircularQueue *queue) {
  queue->front = -1;
  queue->rear = -1;
}
int is_empty(CircularQueue *queue) {
  return queue->front == -1;
}
int is_full(CircularQueue *queue) {
  return(queue->rear + 1) % MAX_SIZE == queue->front;
}
void enqueue(CircularQueue *queue, int item) {
  if(is_full(queue)) {
    printf("queue is full\n");
    return;
  }
  if(is_empty(queue)) {
    queue->front = queue->rear = 0;
  } else {
      queue->rear = (queue->rear + 1) % MAX_SIZE;
  }
  queue->items[queue->rear] = item;
}

int dequeue(CircularQueue *queue) {
  if(is_empty(queue)) {
    printf("queue is empty\n");
    return -1;
  }
  int item = queue->items[queue->front];
  if(queue->front == queue->rear) {
    queue->front = queue->rear = -1;
  } else {
      queue->front = (queue->front + 1) % MAX_SIZE;
  }
  return item;
}

void print_queue(CircularQueue *queue) {
  if(is_empty(queue)) {
    printf("queue is empty");
    return;
  }
  printf("queue: ");
  for(int i = queue->front; i != queue->rear; i = (i + 1) % MAX_SIZE) {
    printf("%d ", queue->items[i]);
  }
  printf("%d\n", queue->items[queue->rear]);
}

int main() {
  CircularQueue queue;
  init_queue(&queue);

  enqueue(&queue, 10);
  enqueue(&queue, 20);
  enqueue(&queue, 30);

  print_queue(&queue);
  printf("dequeued item: %d\n", dequeue(&queue));
  printf("dequeued item: %d\n", dequeue(&queue));

  print_queue(&queue);

  return 0;
}

File: python/circular_queue.py
class CircularQueue:
  def __init__(self, size):
    self.size = size
    self.queue = [None] * size
    self.front = self.rear = -1

  def is_empty(self):
    return self.front == -1

  def is_full(self):
    return (self.rear + 1) % self.size == self.front

  def enqueue(self, item):
    if self.is_full():
      print("queue is full")
      return
    if self.is_empty():
      self.front = self.rear = 0
    else:
      self.rear = (self.rear + 1) % self.size
    self.queue[self.rear] = item

  def dequeue(self):
    if self.is_empty():
      print("queue is empty")
      return
    item = self.queue[self.front]
    if self.front == self.rear:
      self.front = self.rear = -1
    else:
      self.front = (self.front + 1) % self.size
    return item

  def print_queue(self):
    if self.is_empty():
      print("queue is empty")
      return
    print("queue: ")
    for i in range(self.front, self.rear + 1):
      print(self.queue[i % self.size], end="")
    print()

